src/client/pfs_v2.go

```diff
@@ -50,57 +51,72 @@ func (c APIClient) NewPutTarClient(repo, commit string) (_ *PutTarClient, retErr
 // PutTar puts a tar stream into PFS.
 // The files are not persisted until the PutTarClient is closed.
 func (ptc *PutTarClient) PutTar(r io.Reader, tag ...string) error {
-	if len(tag) > 0 {
-		if len(tag) > 1 {
-			return errors.Errorf("PutTar called with %v tags, expected 0 or 1", len(tag))
+	return ptc.maybeError(func() error {
+		if len(tag) > 0 {
+			if len(tag) > 1 {
+				return errors.Errorf("PutTar called with %v tags, expected 0 or 1", len(tag))
+			}
+			if err := ptc.client.Send(&pfs.PutTarRequestV2{Tag: tag[0]}); err != nil {
+				return err
+			}
 		}
-		if err := ptc.client.Send(&pfs.PutTarRequest{Tag: tag[0]}); err != nil {
+		if _, err := grpcutil.ChunkReader(r, func(data []byte) error {
+			return ptc.client.Send(&pfs.PutTarRequestV2{Data: data})
+		}); err != nil {
 			return err
 		}
+		return ptc.client.Send(&pfs.PutTarRequestV2{EOF: true})
+	})
+}
+
+func (ptc *PutTarClient) maybeError(f func() error) (retErr error) {
```

Comment:
This method seems kind of wonky to me, I generally think `maybe` methods are a little lame, since every method that returns an error is implicitly assumed to only maybe do the thing it's supposed to do. It also seems really weird to use that defer to to call `ScrubGRPC` on the error, when it looks like the error can only come from the call to `f`. `ScrubGRPC` is also setup to handle nil errors, and be idempotent. So I think you can just do `return ScrubeGRPC(f())` and remove the defer to get the same result. But I'm wondering if maybe there's a way to restructure this code to not need the helper method at all.

Reply:
The general idea is that it handles a client in a bad state, and will report the initial error on subsequent calls. We don't want the functions to run if the client is in a bad state because we don't have any recovery mechanisms (basically we don't know what got uploaded and what didn't). Also, the pattern of scrubbing in the defer is a good decision from a maintenance standpoint because we don't have to make changes later if more code gets added. A helper function allows us to encapsulate this logic (immediately return the error if the client is in a bad state, and scrub grpc errors) in one place, which seems like a good design. The function could probably be renamed though.

Reply:
The closure is definitely tricky to follow.  I know an if + defer to assign to `ptc.err` is more code per use, but I think it's easier to read.

Reply:
I think I am going to leave this as is for now. I think we should give this pattern a chance since it encapsulates the handling of error state in one place. If it still seems wonky over time, we can easily switch to something else.